Monitor ENF calculation more closely for better progress feedback (EXPOSUREAPP-2743)

d4rken opened the pull request:
## Why
We want to display better progress feedback, we want to know when the download is happening and when we are waiting for the ENF to check the keys.

The goal is to show two different progress types, this PR does not yet add that, but adds the necessary base classes. For now the `isRefreshing` state will be displayed on the risk card if either the transaction is on-going, or the ENF is still calculating.

## How
This PR adds the `CalculationTracker`, which is a set of classes that can track the ENF calcluations that happen outside of our app process.
The idea is:
* Every time we submit submit key files to the ENF, we store an identifier (the token we also pass to the ENF).
* Register for both `ACTION_EXPOSURE_STATE_UPDATED` and `ACTION_EXPOSURE_NOT_FOUND`
* When we receive a broadcast for either action, the token is included returned to us by the ENF and we mark the respective entry as finished.

To the rest of the app we expose a `CalculationTracker.calculations: Flow<Map<String, Calculation>>`:
* We store the last 5 started calculations.
* If any of those does not have a `finishedAt` timestamp set by us, it's still calculating.
* We consider the calculation still in progress as long as there is at least one unfinished calculation.
* The data should be accessed via `ENFClient` which offers easier access with
    * `isCurrentlyCalculating(): Flow<Boolean>`
    * `latestFinishedCalculation(): Flow<Calculation?>`

To notice any issues within the ENF that don't throw us an exception, there is a timeout checker running within the calculation tracker. 
Every 5 minutes it will evaluate the `startedAt` timestamps of all known calculations. If it exceeds the timeout (currently 60 minutes,  @mlenkeit thoughts on that?, might be a bit high), the calculation is considered "timed out", get the result type "TIMEOUT", and given the current time as `finishedAt` timestamp (which marks it as no longer "in calculation").

The calculation data is persisted everytime there is an update to the data map. The class `CalculationTrackerStorage` loads and stores the map as serialized JSON file to our private app storage.

## Testing
* Observe the log for `CalculationTracker`
* Start risk level calculations and observe the log and the progress state of the risk cards.

d4rken committed:
Add a `CalcluationTracker` class that can tell us about the ENF's calcluations.

* Introduces new reactive data provider "HotData" that allows lazy init and safe updates.

d4rken committed:
Add calculation time out enforcer (60min), check every 5min.

d4rken committed:
Show `isRefreshing` when we are either downloading keys or the ENF is still calculating.

d4rken committed:
Rename "token" to "identifier", we'll likely keep using this class for ENF Window Mode,
which has no more "tokens".

d4rken committed:
Remove "state", we can only know whether it's running and that can be based on the `finishedAt` timestamp.
Also improve test readability a bit and added more edge cases when checking timeouts.

d4rken committed:
Additional test cases for ENF calculation edge cases

d4rken committed:
We are fine with tracking the last 5 calculations.

d4rken committed:
Exclude timeouts from `latestFinishedCalculation`

d4rken commented:
In another PR we should consider replacing the timestamp on the risk card, which currently reflects the last time the download was finished, but it makes more sense if it's the timestamp of the newest tracked calculation (which would IMHO be the fix for #948 .

d4rken committed:
HotData should be named HotDataFlow

d4rken committed:
Clean up coroutine/flow packages.

d4rken committed:
Remove additional combineTransform instances, to prevent accidental casting errors in the future.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/flow/HotDataFlow.kt:
```diff
@@ -0,0 +1,68 @@
+import timber.log.Timber
+import kotlin.coroutines.CoroutineContext
+
+class HotDataFlow<T : Any>(
```
This PR also introduces a nice helper class dubbed `HotDataFlow`.

It's a flowable that
* Does lazy init. The start value is provided via a suspending function that is only executed if someone subscribes
* Shares all updates between all subscribers as long as there is at least one subscriber.
* Allows threadsafe updates to it's contents as it only allows its data to be changed by giving it an "update function" that takes the current value and can return a new one.

ralfgehrer replied:
Turned out very nicely.

d4rken committed:
We should only forward calls to the provider and the calculation tracker if the list of key files is non-empty.

d4rken committed:
Add unit test to check for GSON data class restoration behavior with transient fields.
GSON sets it to false, and does not eval the properties to set them.

d4rken committed:
Lints, Lints, Lints

d4rken committed:
Only check the calculation status of the newest submission.
The chance for overlapping calculations is rare with batched key submission,
and if gives us a lower chance of actually being affected by timeouts.

d4rken committed:
Merge branch 'release/1.6.x' into feature/2743-better-progress-feedback

d4rken commented:
I've changed the `ENF.isCalculating` to only check the state of the newest submission. With batched submission there is only a small chance for overlapping calculations and it reduces the chance that we actually run into a timeout and the user sees a prolonged "progress" state. As we currently submit all key files each time, each new calculation would superseed the previous anyways.

vaubaehn commented:
You [invited for feedback](https://github.com/corona-warn-app/cwa-app-android/issues/948#issuecomment-715325151). First class. Nothing else to say. 👍

mlenkeit commented:
@d4rken as discussed, let's set the timeout to 15 minutes initially, should already be quite conservative.

ralfgehrer commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/calculationtracker/CalculationTrackerStorage.kt:
```diff
@@ -0,0 +1,61 @@
+        if (storageDir.mkdirs()) {
+            Timber.v("Created %s", storageDir)
+        }
+        GsonBuilder().create()
```
I'm surprised, not injected :) ?

d4rken replied:
:fearful:

d4rken committed:
Reduce timeout for tracked calculations to 15 minutes after discussion with Maximilian.
In worst case scenarios no calculation exceeded 7 minutes.

d4rken committed:
Reduce timeout check interval to 3 minutes, due to lowered overall timeout limit.

d4rken committed:
Add additional test case that checks that a late result, past timeout, overwrites the timeout.

d4rken committed:
Create `BaseGson` instance within the DI graph use it.
Allows use to later set global settings (pretty print for testers?),
or hook up custom serializers app-wide.

d4rken committed:
Merge branch 'release/1.6.x' into feature/2743-better-progress-feedback

BMItter commented:
passed deep look & test. lgtm
